### ISSUE TITLE
Camera feed not loading on macOS Catalina

### ISSUE THREAD

**Issue by charlielee** [charlielee/boats-animator#253]
The camera feed does not seem to be working on macOS 10.14 or 10.15. A bit of a major issue!

I think the issue is that the camera permission request dialogue is not triggered. See:
https://developer.apple.com/documentation/avfoundation/cameras_and_media_capture/requesting_authorization_for_media_capture_on_macos

Possibly some solutions listed here:
https://stackoverflow.com/questions/56084303/opencv-command-line-app-cant-access-camera-under-macos-mojave

Probably need to mess about with the generated plist file. Probably something to handle within https://github.com/charlielee/nwjs-packager

**Comment by charlielee** [comment]
Electron handles this nicely it seems:
https://electronjs.org/docs/api/system-preferences#systempreferencesaskformediaaccessmediatype-macos

A related issue:
https://github.com/electron/electron/issues/15985

I'm struggling to find anything similar for nodejs

**Comment by charlielee** [comment]
Maybe related to https://github.com/nwjs/nw.js/issues/7229

**Comment by charlielee** [comment]
Maybe this'll help https://github.com/gch1p/mojave-permissions

**Comment by arudnev** [comment]
I just stumbled upon this ticket when I realized that we had similar issue for a while.
Not sure how it worked for us on macOS Mojave, but after we added notarization support for app to work on macOS Catalina we noticed that camera and microphone permissions are not requested.
Adding missing entitlements as part of codesign process seems to resolve the issue.

https://github.com/nwjs/nw.js/issues/7330

So, something like that as entitlements file should fix it

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>com.apple.security.device.camera</key>
    <true/>
    <key>com.apple.security.device.audio-input</key>
    <true/>
    <key>com.apple.security.device.microphone</key>
    <true/>
    <key>com.apple.security.automation.apple-events</key>
    <true/>
    <key>com.apple.security.cs.allow-dyld-environment-variables</key>
    <true/>
    <key>com.apple.security.cs.allow-jit</key>
    <true/>
    <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
    <true/>
    <key>com.apple.security.cs.disable-executable-page-protection</key>
    <true/>
    <key>com.apple.security.cs.disable-library-validation</key>
    <true/>
  </dict>
</plist>
```

**Comment by charlielee** [comment]
@arudnev thank you for your advice! Much appreciated. Boats Animator is not a signed application so there isn't a codesign process as such.

Would the solution to create a `Entitlements.plist` file in package?

**Comment by charlielee** [comment]
I've found the issue and reported it to nw-builder here: https://github.com/nwjs-community/nw-builder/issues/524

**Comment by charlielee** [comment]
Issues like this really do make me consider moving this project to Electron.  nw-builder isn't very maintained unfortunately and NW.js doesn't really have  supported alternative. A lot of functionality that would simply be available in https://github.com/electron/electron-packager I am having to write myself in https://github.com/charlielee/nwjs-packager

**Comment by arudnev** [comment]
Yeah, we've been using our fork (mostly signing, notarization, tune up of windows installer) of https://github.com/evshiron/nwjs-builder-phoenix, which is currently nwjs recommended packager.

**Comment by charlielee** [comment]
This should be fixed in the exportVideo branch since I have extensively rewritten https://github.com/charlielee/nwjs-packager to allow the ffmpeg dependency to be included in this. This rewrite removes the nw-builder dependency.

**Comment by charlielee** [comment]
Fixed in #258
